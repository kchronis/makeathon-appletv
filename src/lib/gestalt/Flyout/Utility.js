// @flow
import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

export type Size = {
  width: number,
  height: number,
};

export type Point = {
  x: number,
  y: number,
};

export type Rect = {
  x: number,
  y: number,
  width: number,
  height: number,
};

export type FlyoutPlacement = 'left' | 'right' | 'top' | 'bottom' | 'auto';

export type ContainerStyle = {
  top: number,
  left: number,
  maxWidth: number,
  maxHeight: number,
};

export const capitalizeFirstLetter = (string?: string): string =>
  !string ? '' : string.charAt(0).toUpperCase() + string.slice(1);

// TODO: Figure out how to flow type that better. The input parameter is a
// StyleSheet object
export const findDirectionWithoutColor = (arrowStyle: any): string => {
  const directionsWithColor = Object.keys(arrowStyle || {})
    .filter(key => /border[a-z]*Color/i.test(key))
    .filter(key => arrowStyle[key] === undefined)
    .map(key =>
      key
        .replace(/border/i, '')
        .replace(/color/i, '')
        .toLowerCase()
    );

  return directionsWithColor[0];
};

export const getArrowStyle = ({
  anchorPoint,
  origin,
  arrowSize,
  placement,
}: {
  anchorPoint: Point,
  origin: Point,
  arrowSize: Size,
  placement: FlyoutPlacement,
}): StyleObj => {
  let width;
  let height;
  // swap width and height for left / right positioned tooltips
  if (placement === 'top' || placement === 'bottom') {
    width = arrowSize.width;
    height = arrowSize.height * 2;
  } else {
    width = arrowSize.height * 2;
    height = arrowSize.width;
  }

  return {
    width,
    height,
    left: placement === 'right' ? 0 : anchorPoint.x - origin.x - width / 2.0,
    top: placement === 'bottom' ? 0 : anchorPoint.y - origin.y - height / 2.0,
    borderTopWidth: height / 2.0,
    borderRightWidth: width / 2.0,
    borderBottomWidth: height / 2.0,
    borderLeftWidth: width / 2.0,
    borderTopColor: placement === 'top' ? undefined : 'transparent',
    borderRightColor: placement === 'right' ? undefined : 'transparent',
    borderBottomColor: placement === 'bottom' ? undefined : 'transparent',
    borderLeftColor: placement === 'left' ? undefined : 'transparent',
  };
};
