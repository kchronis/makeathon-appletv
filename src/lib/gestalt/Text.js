/**
 * WIP - Gestalt like Text
 * https://github.com/pinterest/gestalt/blob/master/src/Text/
 * @flow
 *
 * We are using explicitly a class component to be able to get the native Props
 * callback and pass them through. Otherwise it would not be possible to wrap
 * the Text into any Touchable: https://facebook.github.io/react-native/docs/direct-manipulation.html
 *
 */

import * as React from 'react';
import { Text as ReactText, View, StyleSheet } from 'react-native';
import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';
// https://github.com/facebook/react-native/issues/15955
import type { NativeMethodsMixinType } from 'react-native/Libraries/Renderer/shims/ReactNativeTypes';

import PIColor, { gestaltColorToColor } from '../styles/picolor';
import PIText from '../styles/pitext';
import { isSmallPhoneDevice } from '../Utils';

const textSizeToPITextSize = {
  xsText: PIText.textExtraSmallFontSize,
  smText: PIText.textSmallFontSize,
  mdText: PIText.textMediumFontSize,
  lgText: PIText.textLargeFontSize,
  xlText: PIText.textExtraLargeFontSize,
  xsDisplay: PIText.displayExtraSmallFontSize,
  smDisplay: PIText.displayExtraSmallFontSize,
  mdDisplay: PIText.displayMediumFontSize,
  lgDisplay: PIText.displayLargeFontSize,
  xlDisplay: PIText.displayExtraLargeFontSize,
};

const textSmallPhoneSizeToPITextSize = {
  xsText: PIText.textSmallPhoneExtraSmallFontSize,
  smText: PIText.textSmallPhoneSmallFontSize,
  mdText: PIText.textSmallPhoneMediumFontSize,
  lgText: PIText.textSmallPhoneLargeFontSize,
  xlText: PIText.textSmallPhoneExtraLargeFontSize,
  xsDisplay: PIText.displaySmallPhoneExtraSmallFontSize,
  smDisplay: PIText.displaySmallPhoneExtraSmallFontSize,
  mdDisplay: PIText.displaySmallPhoneMediumFontSize,
  lgDisplay: PIText.displaySmallPhoneLargeFontSize,
  xlDisplay: PIText.displaySmallPhoneExtraLargeFontSize,
};

const getTextSizeToPITextSize = (textSize: string) =>
  isSmallPhoneDevice()
    ? textSmallPhoneSizeToPITextSize[textSize]
    : textSizeToPITextSize[textSize];

type Props = {|
  align?: 'auto' | 'left' | 'right' | 'center' | 'justify',
  bold?: boolean,
  children?: React.Node,
  color?:
    | 'clear'
    | 'primaryBackground'
    | 'darkGray'
    | 'lightGray'
    | 'superLightGray'
    | 'red'
    | 'elephantPinBackground'
    | 'white',
  italic?: boolean,
  size?:
    | 'xsText'
    | 'smText'
    | 'mdText'
    | 'lgText'
    | 'xlText'
    | 'xsDisplay'
    | 'smDisplay'
    | 'mdDisplay'
    | 'lgDisplay'
    | 'xlDisplay',
  accessible?: boolean,
  testID?: string,
  onPress?: () => void,
  onLongPress?: () => void,
  suppressHighlighting?: boolean,

  style?: StyleObj, // React Native only: Should go away eventually
|};

export default class Text extends React.PureComponent<Props> {
  _root: ?NativeMethodsMixinType;

  setNativeProps(nativeProps: Object) {
    if (this._root) {
      this._root.setNativeProps(nativeProps);
    }
  }

  render() {
    const {
      align = 'left',
      bold = false,
      children,
      color = 'darkGray',
      italic = false,
      size = 'mdText',

      accessible = true, // React Native only
      testID, // React Native only
      onPress, // React Native only
      onLongPress, // React Native only
      suppressHighlighting, // React Native only

      style, // React Native only
    } = this.props;
    const textStyles = {
      textAlign: align,
      fontFamily: PIText.defaultFont,
      fontWeight: bold ? 'bold' : 'normal',
      fontStyle: italic ? 'italic' : 'normal',
      fontSize: getTextSizeToPITextSize(size),
      color: gestaltColorToColor[color],
    };
    /* eslint-disable no-return-assign */
    return (
      <ReactText
        ref={root => (this._root = root)}
        style={StyleSheet.flatten([textStyles, style])}
        accessible={accessible}
        testID={testID}
        onPress={onPress}
        onLongPress={onLongPress}
        suppressHighlighting={suppressHighlighting}
      >
        {children}
      </ReactText>
    );
    /* eslint-enable no-return-assign */
  }
}
