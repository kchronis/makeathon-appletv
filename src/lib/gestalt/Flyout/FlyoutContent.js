// @flow

import * as React from 'react';
import {
  StyleSheet,
  View,
  Dimensions,
  NativeModules,
  findNodeHandle,
} from 'react-native';
import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';
import type { LayoutEvent } from './../CoreEventTypes';
import computeGeometry from './Geometry';
import { capitalizeFirstLetter, findDirectionWithoutColor } from './Utility';
import type {
  Size,
  Point,
  Rect,
  FlyoutPlacement,
  ContainerStyle,
} from './Utility';

const invariant = require('fbjs/lib/invariant');

type UIManagerType = {
  measure: (
    ref: ?React.Node,
    (
      x: number,
      y: number,
      width: number,
      height: number,
      pageX: number,
      pageY: number
    ) => void
  ) => void,
};

const UIManager = (NativeModules.UIManager: UIManagerType);

const {
  height: SCREEN_HEIGHT,
  width: SCREEN_WIDTH,
}: { height: number, width: number } = Dimensions.get('window');

const MAX_Z_INDEX = 9999;

type Props = {
  containerSize: Size, // The available size for the FlyoutContent
  fromRect: ?Rect, // Rect that the FlyoutContent will attach to
  fromView: ?React.ElementRef<typeof React.Component>, // Ref for the element that the FlyoutContent will attach to
  padding: number, // The padding for the FlyoutContent content
  arrowColor: string, // The color of the arrow
  arrowWidth: number, // The width of the arrow
  arrowHeight: number, // The height of the arrow
  placement: FlyoutPlacement, // The placeement of the FlyoutContent
  pointerEvents: string, // Controls whether the FlyoutContent can be the target of touch events.
  offset: Point, // Define an offset to the position calculated by the FlyoutContent
  component: () => React.Node, // Component to render in the FlyoutContent
};

type State = {
  containerStyle: ContainerStyle,
  arrowStyle: ?StyleObj,
  containerPaddingForArrowStyle: ?StyleObj,
  width: number,
  height: number,
};

export default class FlyoutContent extends React.Component<Props, State> {
  static defaultProps = {
    // For now we assume we have the whole screen width and height to show the
    // Flyout
    containerSize: {
      width: SCREEN_WIDTH,
      height: SCREEN_HEIGHT,
    },
    fromRect: null,
    fromView: null,
    pointerEvents: 'box-none',
    offset: {
      x: 0,
      y: 0,
    },
  };

  constructor(props: Props) {
    super(props);

    const { containerSize, padding } = props;
    this.state = {
      // Render offscren initially
      containerStyle: {
        top: -9999,
        left: -9999,
        maxWidth: containerSize.width - padding * 2,
        maxHeight: containerSize.height - padding * 2,
      },
      arrowStyle: null,
      containerPaddingForArrowStyle: null,
      width: 0,
      height: 0,
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.props === nextProps) {
      return;
    }

    this._measureAndComputeStylesForProps(nextProps);
  }

  render() {
    const { pointerEvents, component, arrowColor } = this.props;
    const {
      containerStyle,
      arrowStyle,
      containerPaddingForArrowStyle,
    } = this.state;
    const borderDirection = capitalizeFirstLetter(
      findDirectionWithoutColor(arrowStyle)
    );

    return (
      <View
        style={StyleSheet.flatten([
          styles.absolute,
          containerStyle,
          containerPaddingForArrowStyle,
          { zIndex: MAX_Z_INDEX },
        ])}
      >
        <View onLayout={this._onLayout}>{React.createElement(component)}</View>
        <View
          style={[
            styles.absolute,
            arrowStyle,
            { [`border${borderDirection}Color`]: arrowColor },
          ]}
          pointerEvents={pointerEvents}
        />
      </View>
    );
  }

  _onLayout = (event: LayoutEvent) => {
    const { width, height } = event.nativeEvent.layout;

    this.setState({ width, height }, () => {
      this._measureAndComputeStylesForProps(this.props);
    });
  };

  _measureAndComputeStylesForProps = (props: Props) => {
    // For some reason if we use props.fromView in here flow does not
    // detect that we use the fromView prop
    const { fromView } = props;
    if (fromView) {
      // Try to measure the fromView and than compute the style
      // We use the fromView to attach the Flyout automatically
      UIManager.measure(
        findNodeHandle(fromView),
        (x: number, y: number, fromWidth: number, fromHeight: number) => {
          // Update fromRect to the fromView values
          const newProps = {
            ...props,
            fromRect: {
              x,
              y,
              width: fromWidth,
              height: fromHeight,
            },
          };
          this._computeStyles(newProps);
        }
      );
      return;
    }

    // No fromView given let's use the fromRect passed in
    this._computeStyles(props);
  };

  // Compute the new styles for given props and update the state
  _computeStyles = (props: Props): void => {
    // We are passing in the whole props and destructing it here and not just a
    // subset as flow would not detect that the props we are passing in are
    // actually used
    const {
      arrowHeight,
      arrowWidth,
      containerSize,
      padding,
      placement,
    } = props;
    let { fromRect } = props;
    // Assert in development and just assign a default object in production
    // better maybe just crash there too.
    invariant(fromRect, 'No valid fromView or fromRect was given.');
    // eslint-disable-next-line no-param-reassign
    fromRect = fromRect || { x: 0, y: 0, width: 0, height: 0 };

    const { offset } = this.props;
    const { width, height, containerStyle } = this.state;
    const geometry = computeGeometry({
      contentSize: { width, height },
      fromRect,
      containerSize,
      arrowSize: {
        width: arrowWidth,
        height: arrowHeight,
      },
      padding,
      placement,
    });

    // TODO: How do we handle the case if geometry is undefined
    invariant(geometry, 'No valid geometry could be computed.');

    this.setState({
      containerStyle: {
        ...containerStyle,
        top: geometry.origin.y + offset.y,
        left: geometry.origin.x + offset.x,
      },
      arrowStyle: geometry.arrowStyle,
      containerPaddingForArrowStyle: geometry.containerPaddingForArrowStyle,
    });
  };
}

const styles = StyleSheet.create({
  absolute: {
    position: 'absolute',
  },
});
