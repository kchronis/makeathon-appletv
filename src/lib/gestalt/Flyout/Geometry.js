// @flow

import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import { getArrowStyle } from './Utility';
import type { Size, Point, Rect, FlyoutPlacement } from './Utility';

export type Geometry = {
  origin: Point,
  arrowStyle: StyleObj,
  containerPaddingForArrowStyle: StyleObj,
  placement: FlyoutPlacement,
};

const createPoint = (x: number, y: number): Point => ({ x, y });

// Calculate the geometry based on onput parameters. We don't support automatically
// adjust the alignment if top, bottom, left or right is choosen and there is no space for it
const computeGeometry = ({
  contentSize,
  placement,
  fromRect,
  containerSize,
  arrowSize,
  padding,
}: {
  contentSize: Size,
  placement: FlyoutPlacement,
  fromRect: Rect,
  containerSize: Size,
  arrowSize: Size,
  padding: number,
}): ?Geometry => {
  const displayArea = {
    x: padding,
    y: padding,
    width: containerSize.width - padding * 2,
    height: containerSize.height - padding * 2,
  };

  const options = {
    displayArea,
    fromRect,
    contentSize,
    arrowSize,
  };

  switch (placement) {
    case 'top':
      return computeTopGeometry(options);
    case 'bottom':
      return computeBottomGeometry(options);
    case 'left':
      return computeLeftGeometry(options);
    case 'right':
      return computeRightGeometry(options);
    default:
      return computeAutoGeometry(options);
  }
};

const computeTopGeometry = ({
  displayArea,
  fromRect,
  contentSize,
  arrowSize,
}): Geometry => {
  const placement = 'top';
  const origin = createPoint(
    Math.min(
      displayArea.x + displayArea.width - contentSize.width,
      Math.max(
        displayArea.x,
        fromRect.x + (fromRect.width - contentSize.width) / 2
      )
    ),
    fromRect.y - contentSize.height - arrowSize.height
  );

  const anchorPoint = createPoint(
    fromRect.x + fromRect.width / 2.0,
    fromRect.y
  );
  const arrowStyle = getArrowStyle({
    anchorPoint,
    origin,
    arrowSize,
    placement,
  });
  const containerPaddingForArrowStyle = {
    paddingBottom: arrowStyle.height || 0,
  };

  return {
    origin,
    arrowStyle,
    containerPaddingForArrowStyle,
    placement,
  };
};

const computeBottomGeometry = ({
  displayArea,
  fromRect,
  contentSize,
  arrowSize,
}): Geometry => {
  const placement = 'bottom';
  const origin = createPoint(
    Math.min(
      displayArea.x + displayArea.width - contentSize.width,
      Math.max(
        displayArea.x,
        fromRect.x + (fromRect.width - contentSize.width) / 2
      )
    ),
    fromRect.y + fromRect.height + arrowSize.height
  );

  const anchorPoint = createPoint(
    fromRect.x + fromRect.width / 2.0,
    fromRect.y + fromRect.height
  );
  const arrowStyle = getArrowStyle({
    anchorPoint,
    origin,
    arrowSize,
    placement,
  });
  const containerPaddingForArrowStyle = {
    paddingTop: arrowStyle.height || 0,
  };

  return {
    origin,
    arrowStyle,
    containerPaddingForArrowStyle,
    placement,
  };
};

const computeLeftGeometry = ({
  displayArea,
  fromRect,
  contentSize,
  arrowSize,
}): Geometry => {
  const placement = 'left';
  const origin = createPoint(
    fromRect.x - contentSize.width - arrowSize.height,
    Math.min(
      displayArea.y + displayArea.height - contentSize.height,
      Math.max(
        displayArea.y,
        fromRect.y + (fromRect.height - contentSize.height) / 2
      )
    )
  );

  const anchorPoint = createPoint(
    fromRect.x,
    fromRect.y + fromRect.height / 2.0
  );
  const arrowStyle = getArrowStyle({
    anchorPoint,
    origin,
    arrowSize,
    placement,
  });
  const containerPaddingForArrowStyle = {
    paddingRight: arrowStyle.height || 0,
  };

  return {
    origin,
    arrowStyle,
    containerPaddingForArrowStyle,
    placement,
  };
};

const computeRightGeometry = ({
  displayArea,
  fromRect,
  contentSize,
  arrowSize,
}): Geometry => {
  const placement = 'right';
  const origin = createPoint(
    fromRect.x + fromRect.width + arrowSize.height,
    Math.min(
      displayArea.y + displayArea.height - contentSize.height,
      Math.max(
        displayArea.y,
        fromRect.y + (fromRect.height - contentSize.height) / 2
      )
    )
  );

  const anchorPoint = createPoint(
    fromRect.x + fromRect.width,
    fromRect.y + fromRect.height / 2.0
  );
  const arrowStyle = getArrowStyle({
    anchorPoint,
    origin,
    arrowSize,
    placement,
  });
  const containerPaddingForArrowStyle = {
    paddingLeft: arrowStyle.height || 0,
  };

  return {
    origin,
    arrowStyle,
    containerPaddingForArrowStyle,
    placement,
  };
};

const computeAutoGeometry = ({
  displayArea,
  fromRect,
  contentSize,
  arrowSize,
}): ?Geometry => {
  const placementsToTry = [
    computeLeftGeometry,
    computeRightGeometry,
    computeBottomGeometry,
    computeTopGeometry,
  ];

  let geometry;
  // eslint-disable-next-line no-plusplus
  for (let i = 0; i < placementsToTry.length; i++) {
    const placementFn = placementsToTry[i];
    geometry = placementFn({ displayArea, fromRect, contentSize, arrowSize });
    const { origin } = geometry;
    if (
      origin.x >= displayArea.x &&
      origin.x <= displayArea.x + displayArea.width - contentSize.width &&
      origin.y >= displayArea.y &&
      origin.y <= displayArea.y + displayArea.height - contentSize.height
    ) {
      break;
    }
  }

  return geometry;
};

export default computeGeometry;
