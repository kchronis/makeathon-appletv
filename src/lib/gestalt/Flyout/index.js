/**
 * WIP - Gestalt like Flyout
 * https://github.com/pinterest/gestalt/tree/master/src/Flyout
 * @flow
 */

import * as React from 'react';
// eslint-disable-next-line import/no-duplicates
import { View } from 'react-native';
import FlyoutContent from './FlyoutContent';

import type { Rect } from './Utility';
import { gestaltColorToColor } from '../../styles/picolor';

type Props = {|
  visible: boolean, // The visible prop determines whether your Flyout is visible
  anchor: ?React.ElementRef<typeof React.Component>, // Ref for the element that the Flyout will attach to
  fromRect: ?Rect, // Rect that the FlyoutContent will attach to. This is supposed to go away eventually though
  bgColor: 'darkGray' | 'white' | 'orange', // Background color of Flyout
  children?: any, // Children rendered in the Flyout
  idealDirection: 'top' | 'right' | 'bottom' | 'left' | 'auto', // Preferred direction for the Flyout to open
  // onDismiss: () => void, // Not available on React Native. Use your own TouchableWithoutFeedback
  // positionRelativeToAnchor?: boolean, // Not available on RN

  // TODO: Add support for different sizes
  // size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl', // Widths for the flyout
|};

// const SIZE_WIDTH_MAP = {
//   xs: 185,
//   sm: 230,
//   md: 284,
//   lg: 320,
//   xl: 375,
// };

type State = {};

export default class Flyout extends React.Component<Props, State> {
  static defaultProps = {
    anchor: null,
    fromRect: null,
    bgColor: 'white',
    idealDirection: 'auto',
    // size: 'sm',
  };

  _renderComponent = () => {
    // Calculate the style for the Flyout
    const { bgColor } = this.props;
    const containerStyle = {
      backgroundColor: gestaltColorToColor[bgColor],
      padding: 8,
      borderRadius: 5,
    };
    return <View style={containerStyle}>{this.props.children}</View>;
  };

  render() {
    const { anchor, fromRect, bgColor, idealDirection, visible } = this.props;
    return visible ? (
      <FlyoutContent
        fromView={anchor}
        fromRect={fromRect}
        padding={10}
        placement={idealDirection}
        arrowColor={gestaltColorToColor[bgColor]}
        arrowWidth={16}
        arrowHeight={8}
        component={this._renderComponent}
      />
    ) : null;
  }
}

// Playground support
/* eslint-disable import/first, import/no-duplicates, no-return-assign, react/no-unescaped-entities */
import { Text } from 'react-native';
import Button from './../Button';

// eslint-disable-next-line import/first
import { type PlaygroundDefineType } from '../../../screens/Playground/Playground';

let ref = null;

module.exports.__cards__ = (define: PlaygroundDefineType) => {
  define(
    'Flyover Default',
    (state = { visible: false }, update: mixed => void) => (
      <View
        style={{
          backgroundColor: 'white',
          padding: 10,
          paddingBottom: 50,
        }}
      >
        <Button
          text="Show Flyover"
          color={'red'}
          ref={view => (ref = view)}
          onPress={() => {
            update({ visible: !state.visible });
          }}
        />
        <Flyout visible={state.visible} anchor={ref} bgColor={'orange'}>
          <Text
            style={{
              color: '#fff',
              fontWeight: 'bold',
            }}
          >
            I'm the content of this popover!
          </Text>
        </Flyout>
      </View>
    )
  );
};
