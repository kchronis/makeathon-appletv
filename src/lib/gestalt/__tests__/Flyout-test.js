/* eslint-env jest */
/* eslint-disable react/no-unescaped-entities */
import React from 'react';
import renderer from 'react-test-renderer';

import Flyout from '../Flyout';
import Text from '../Text';

test('Renders Flyout', () => {
  const component = renderer.create(
    <Flyout visible anchor={<Text>test</Text>} bgColor={'orange'}>
      <Text
        style={{
          color: '#fff',
          fontWeight: 'bold',
        }}
      >
        I'm the content of this popover!
      </Text>
    </Flyout>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
