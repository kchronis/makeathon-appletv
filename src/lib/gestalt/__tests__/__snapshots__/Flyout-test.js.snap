// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Flyout 1`] = `
<View
  style={
    Array [
      Object {
        "position": "absolute",
      },
      Object {
        "left": -9999,
        "maxHeight": 1314,
        "maxWidth": 730,
        "top": -9999,
      },
      Object {
        "zIndex": 9999,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
  >
    <View
      style={
        Object {
          "backgroundColor": undefined,
          "borderRadius": 5,
          "padding": 8,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        onLongPress={undefined}
        onPress={undefined}
        style={
          Object {
            "color": "#fff",
            "fontFamily": "HelveticaNeue",
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "bold",
            "textAlign": "left",
          }
        }
        suppressHighlighting={undefined}
        testID={undefined}
      >
        I'm the content of this popover!
      </Text>
    </View>
  </View>
  <View
    pointerEvents="box-none"
    style={
      Array [
        Object {
          "position": "absolute",
        },
        null,
        Object {
          "borderColor": undefined,
        },
      ]
    }
  />
</View>
`;
